executor/write.go

```diff
@@ -886,15 +886,56 @@ func (e *InsertValues) getRows(cols []*table.Column, ignoreErr bool) (rows [][]t
 }
 
 func (e *InsertValues) getRow(cols []*table.Column, list []expression.Expression, ignoreErr bool) ([]types.Datum, error) {
-	vals := make([]types.Datum, len(list))
+	row := make([]types.Datum, len(e.Table.Cols()))
+	hasValue := make([]bool, len(e.Table.Cols()))
+	if err := e.fillDefaultValues(row, ignoreErr); err != nil {
+		return nil, errors.Trace(err)
+	}
+
 	for i, expr := range list {
-		val, err := expr.Eval(nil)
-		vals[i] = val
+		val, err := expr.Eval(row)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		val, err = table.CastValue(e.ctx, val, cols[i].ToInfo())
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		offset := cols[i].Offset
+		row[offset] = val
+		hasValue[offset] = true
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
 	}
-	return e.fillRowData(cols, vals, ignoreErr)
+
+	return e.checkRowData(cols, len(list), hasValue, row, ignoreErr)
+}
+
+func (e *InsertValues) fillDefaultValues(row []types.Datum, ignoreErr bool) error {
+	var defaultValueCols []*table.Column
+	for i, c := range e.Table.Cols() {
+		var err error
+		if c.IsGenerated() {
+			continue
+		} else if mysql.HasAutoIncrementFlag(c.Flag) {
+			row[i] = table.GetZeroValue(c.ToInfo())
+		} else {
+			row[i], err = table.GetColDefaultValue(e.ctx, c.ToInfo())
+			if table.IsNoDefault(err) && mysql.HasNotNullFlag(c.Flag) {
+				row[i] = table.GetZeroValue(c.ToInfo())
+			} else if err != nil {
+				return errors.Trace(err)
+			}
+		}
+		defaultValueCols = append(defaultValueCols, c)
```

Comment:
I think defaultValueCols is not necessary, just use `e.Table.Cols()` in ` table.CastValues(e.ctx, row, e.Table.Cols(), ignoreErr)` ?